File: core/java/src/main/java/hara/lang/protocol/IEncodeVisitor.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.iencodevisitor", name = "IEncodeVisitor")
public interface IEncodeVisitor {
  @HaraMethod(value = "visit-nil", arity = 1) Object visitNil();
  @HaraMethod(value = "visit-boolean", arity = 2) Object visitBoolean(Object value);
  @HaraMethod(value = "visit-number", arity = 2) Object visitNumber(Object value);
  @HaraMethod(value = "visit-character", arity = 2) Object visitCharacter(Object value);
  @HaraMethod(value = "visit-string", arity = 2) Object visitString(Object value);
  @HaraMethod(value = "visit-keyword", arity = 2) Object visitKeyword(Object value);
  @HaraMethod(value = "visit-symbol", arity = 2) Object visitSymbol(Object value);
  @HaraMethod(value = "visit-seq", arity = 2) Object visitSeq(Object value);
  @HaraMethod(value = "visit-vector", arity = 2) Object visitVector(Object value);
  @HaraMethod(value = "visit-map", arity = 2) Object visitMap(Object value);
  @HaraMethod(value = "visit-set", arity = 2) Object visitSet(Object value);
  @HaraMethod(value = "visit-tagged", arity = 3) Object visitTagged(Object tag, Object value);
  @HaraMethod(value = "visit-unknown", arity = 2) Object visitUnknown(Object value);
}

File: core/rust/src/lang/protocol/iencodevisitor.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.iencodevisitor", name = "IEncodeVisitor")]
pub trait IEncodeVisitor {
    type Value;
    type Output;

    #[hara_method(value = "visit-nil", arity = 1)]
    fn visit_nil(&mut self) -> Self::Output;
    #[hara_method(value = "visit-boolean", arity = 2)]
    fn visit_boolean(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-number", arity = 2)]
    fn visit_number(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-character", arity = 2)]
    fn visit_character(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-string", arity = 2)]
    fn visit_string(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-keyword", arity = 2)]
    fn visit_keyword(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-symbol", arity = 2)]
    fn visit_symbol(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-seq", arity = 2)]
    fn visit_seq(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-vector", arity = 2)]
    fn visit_vector(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-map", arity = 2)]
    fn visit_map(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-set", arity = 2)]
    fn visit_set(&mut self, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-tagged", arity = 3)]
    fn visit_tagged(&mut self, tag: Self::Value, value: Self::Value) -> Self::Output;
    #[hara_method(value = "visit-unknown", arity = 2)]
    fn visit_unknown(&mut self, value: Self::Value) -> Self::Output;
}
